discovery/store.go

```diff
@@ -303,29 +329,44 @@ func (s *sqlStore) search(serviceID string, query map[string]string) ([]vc.Verif
 	return results, nil
 }
 
-// updateTimestamp updates the timestamp of the given service.
-// Clients should pass the timestamp they received from the server (which simply sets it).
-// Servers should pass nil (since they "own" the timestamp), which causes it to be incremented.
-func (s *sqlStore) updateTimestamp(tx *gorm.DB, serviceID string, newTimestamp *Timestamp) (Timestamp, error) {
-	var result serviceRecord
+// updateTag updates the tag of the given service.
+// Clients should pass the tag they received from the server (which simply sets it).
+// Servers should pass nil (since they "own" the tag), which causes it to be incremented.
+// It returns
+func (s *sqlStore) updateTag(tx *gorm.DB, serviceID string, newTimestamp *Tag) (*Timestamp, error) {
+	var service serviceRecord
 	// Lock (SELECT FOR UPDATE) discovery_service row to prevent concurrent updates to the same list, which could mess up the lamport timestamp.
 	if err := tx.Clauses(clause.Locking{Strength: "UPDATE"}).
 		Where(serviceRecord{ID: serviceID}).
-		Find(&result).
+		Find(&service).
 		Error; err != nil {
-		return 0, err
+		return nil, err
 	}
-	result.ID = serviceID
+	service.ID = serviceID
+	var result *Timestamp
 	if newTimestamp == nil {
-		// Increment timestamp
-		result.LamportTimestamp++
+		// Update tag: decode current timestamp, increment it, encode it again.
+		currTimestamp := Timestamp(0)
+		if service.LastTag != "" {
+			// If LastTag is empty, it means the service was just created and no presentations were added yet.
+			ts := service.LastTag.Timestamp(service.TagPrefix)
+			if ts == nil {
+				// would be very weird
+				return nil, fmt.Errorf("invalid tag '%s'", service.LastTag)
```

Comment:
Better then to make the columns `not null`?

Reply:
It's not that the tag is null, it can't decode its own tag. Indicates invalid data in SQL (maybe when becoming server, after being client first for a particular service).

I didn't want to think about supporting that at first, since it seemed quite complex. But now tags are server-specific,  it has become managable but requires some SQL-fu. I'll add an issue to add the feature; https://github.com/nuts-foundation/nuts-node/issues/2656

Reply:
I meant that someone could change the values to null in the DB and then that error will be triggered.

Reply:
Or changed the value of one of `service.last_tag` or `service.tag_prefix`, making them not correlate anymore. Making the column non-null won't change that, but I'll make the error message more descriptive.